File: examples/python/zmqcat.py
"""Tiny zmqcat client. Talks to the local sidecar (unix or tcp).

    from zmqcat import Client
    c = Client()          # unix:///tmp/zmqcat-<uid>.sock
    c.put("jobs", "run")
    print(c.take("jobs"))
"""

from __future__ import annotations

import json
import os
import socket
import struct
from typing import Any


MAGIC = b"ZMQC"


def default_listen() -> str:
    return f"unix:///tmp/zmqcat-{os.getuid()}.sock"


def _split(listen: str) -> tuple[str, str]:
    if listen.startswith("unix://"):
        return "unix", listen[len("unix://") :]
    if listen.startswith("tcp://"):
        return "tcp", listen[len("tcp://") :]
    if listen.startswith("/") or listen.startswith("."):
        return "unix", listen
    return "tcp", listen


class Client:
    def __init__(self, listen: str | None = None, name: str = "python"):
        listen = listen or os.environ.get("ZMQCAT_LISTEN") or default_listen()
        family, address = _split(listen)
        if family == "unix":
            self.sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
            self.sock.connect(address)
        else:
            host, port = address.rsplit(":", 1)
            self.sock = socket.create_connection((host, int(port)))
        self.name = name
        self.hello(name)

    def close(self) -> None:
        self.sock.close()

    def hello(self, name: str) -> dict[str, Any]:
        return self._rpc({"op": "hello", "from": name})

    def put(self, mailbox: str, text: str = "", body: bytes | None = None) -> dict[str, Any]:
        return self._rpc(self._payload("put", mailbox, text, body))

    def take(self, mailbox: str) -> dict[str, Any]:
        return self._rpc({"op": "take", "name": mailbox, "from": self.name})

    def pub(self, topic: str, text: str = "", body: bytes | None = None) -> dict[str, Any]:
        return self._rpc(self._payload("pub", topic, text, body))

    def sub(self, prefix: str = "") -> dict[str, Any]:
        return self._rpc({"op": "sub", "name": prefix, "from": self.name})

    def recv(self) -> dict[str, Any]:
        return self._read_skip_ping()

    def ping(self) -> dict[str, Any]:
        return self._rpc({"op": "ping"})

    def ready(self, service: str) -> dict[str, Any]:
        return self._rpc({"op": "ready", "name": service, "from": self.name})

    def req(self, service: str, text: str = "", body: bytes | None = None, timeout: float = 1.0, attempts: int = 3) -> dict[str, Any]:
        frame = self._payload("req", service, text, body)
        if "id" not in frame:
            import time

            frame["id"] = f"r-{time.time_ns()}"
        last: Exception | None = None
        for _ in range(max(1, attempts)):
            try:
                self.sock.settimeout(timeout)
                out = self._rpc(frame)
                self.sock.settimeout(None)
                return out
            except Exception as e:
                last = e
                self.sock.settimeout(None)
        raise RuntimeError(f"request abandoned after {attempts} attempts: {last}")

    def rep(self, corr_id: str, text: str = "", name: str = "", body: bytes | None = None) -> dict[str, Any]:
        f = self._payload("rep", name, text, body)
        f["id"] = corr_id
        return self._rpc(f)

    def reserve(self, mailbox: str, lease: int = 60) -> dict[str, Any]:
        return self._rpc({"op": "reserve", "name": mailbox, "from": self.name, "lease": lease})

    def ack(self, delivery: str) -> dict[str, Any]:
        return self._rpc({"op": "ack", "delivery": delivery})

    def nack(self, delivery: str) -> dict[str, Any]:
        return self._rpc({"op": "nack", "delivery": delivery})

    def _payload(self, op: str, name: str, text: str, body: bytes | None) -> dict[str, Any]:
        f: dict[str, Any] = {"op": op, "name": name, "from": self.name, "text": text}
        if body:
            import base64

            f["body"] = base64.b64encode(body).decode("ascii")
        return f

    def _rpc(self, f: dict[str, Any]) -> dict[str, Any]:
        self._write(f)
        out = self._read_skip_ping()
        if out.get("op") == "err":
            raise RuntimeError(out.get("error") or "zmqcat error")
        return out

    def _read_skip_ping(self) -> dict[str, Any]:
        while True:
            out = self._read()
            if out.get("op") == "ping":
                self._write({"op": "pong", "id": out.get("id", "")})
                continue
            return out

    def _write(self, f: dict[str, Any]) -> None:
        raw = json.dumps(f).encode()
        self.sock.sendall(MAGIC + struct.pack(">I", len(raw)) + raw)

    def _read(self) -> dict[str, Any]:
        hdr = _readn(self.sock, 8)
        if hdr[:4] != MAGIC:
            raise RuntimeError(f"bad magic {hdr[:4]!r}")
        (n,) = struct.unpack(">I", hdr[4:])
        raw = _readn(self.sock, n)
        return json.loads(raw)


def _readn(sock: socket.socket, n: int) -> bytes:
    buf = bytearray()
    while len(buf) < n:
        chunk = sock.recv(n - len(buf))
        if not chunk:
            raise EOFError("zmqcat socket closed")
        buf.extend(chunk)
    return bytes(buf)


if __name__ == "__main__":
    import sys

    c = Client()
    if len(sys.argv) < 2:
        print(c.ping())
        sys.exit(0)
    op = sys.argv[1]
    if op == "put":
        print(c.put(sys.argv[2], " ".join(sys.argv[3:])))
    elif op == "take":
        print(c.take(sys.argv[2]))
    elif op == "pub":
        print(c.pub(sys.argv[2], " ".join(sys.argv[3:])))
    elif op == "sub":
        c.sub(sys.argv[2] if len(sys.argv) > 2 else "")
        while True:
            print(c.recv())
    elif op == "ready":
        print(c.ready(sys.argv[2]))
    elif op == "req":
        print(c.req(sys.argv[2], " ".join(sys.argv[3:])))
    else:
        raise SystemExit("put|take|pub|sub|ready|req")

File: examples/openresty/zmqcat.lua
-- resty/zmqcat.lua — OpenResty client for a local zmqcat sidecar.
--
--     local zmqcat = require "resty.zmqcat"
--     local c, err = zmqcat.connect()
--     c:put("jobs", '{"run":true}')
--     local msg, err = c:take("jobs")
--
-- Default socket: unix:/tmp/zmqcat-<uid>.sock  (set ZMQCAT_LISTEN)

local cjson = require "cjson.safe"

local _M = { _VERSION = "0.1.0" }
local mt = { __index = _M }

local MAGIC = "ZMQC"

local function be32(n)
  local b1 = math.floor(n / 16777216) % 256
  local b2 = math.floor(n / 65536) % 256
  local b3 = math.floor(n / 256) % 256
  local b4 = n % 256
  return string.char(b1, b2, b3, b4)
end

local function un_be32(s)
  local b1, b2, b3, b4 = s:byte(1, 4)
  return b1 * 16777216 + b2 * 65536 + b3 * 256 + b4
end

local function default_listen()
  local env = os.getenv("ZMQCAT_LISTEN")
  if env and env ~= "" then
    return env
  end
  -- ngx.worker.pid is not uid; unix path is owned by the serve process user.
  return "unix:/tmp/zmqcat-" .. (os.getenv("UID") or "0") .. ".sock"
end

local function parse(listen)
  listen = listen or default_listen()
  if listen:sub(1, 7) == "unix://" then
    return "unix:" .. listen:sub(8)
  end
  if listen:sub(1, 6) == "tcp://" then
    return listen:sub(7)
  end
  if listen:sub(1, 1) == "/" then
    return "unix:" .. listen
  end
  if listen:sub(1, 5) == "unix:" then
    return listen
  end
  return listen
end

local function write_frame(sock, obj)
  local raw = cjson.encode(obj)
  if not raw then
    return nil, "encode"
  end
  return sock:send(MAGIC .. be32(#raw) .. raw)
end

local function read_frame(sock)
  local hdr, err = sock:receive(8)
  if not hdr then
    return nil, err
  end
  if hdr:sub(1, 4) ~= MAGIC then
    return nil, "bad magic"
  end
  local n = un_be32(hdr:sub(5, 8))
  if n <= 0 or n > 2 * 1024 * 1024 then
    return nil, "bad length"
  end
  local raw, err2 = sock:receive(n)
  if not raw then
    return nil, err2
  end
  local obj, jerr = cjson.decode(raw)
  if not obj then
    return nil, jerr
  end
  return obj
end

function _M.connect(listen, name)
  local sock = ngx.socket.tcp()
  sock:settimeouts(1000, 60000, 60000)
  local addr = parse(listen)
  local ok, err = sock:connect(addr)
  if not ok then
    return nil, err
  end
  -- Correlation ids double as mailbox message ids, so they must be unique
  -- across clients: a per-connection counter alone would let two workers
  -- deduplicate each other's requests.
  local prefix = string.format("%08x%08x", math.random(0, 0xffffffff), ngx.now() * 1000 % 0xffffffff)
  local self = setmetatable({ sock = sock, name = name or "openresty", prefix = prefix, seq = 0 }, mt)
  local f, herr = self:hello(self.name)
  if not f then
    sock:close()
    return nil, herr
  end
  return self
end

function _M:hello(name)
  return self:rpc({ op = "hello", from = name or self.name })
end

function _M:put(mailbox, text)
  return self:rpc({ op = "put", name = mailbox, from = self.name, text = text or "" })
end

function _M:take(mailbox)
  return self:rpc({ op = "take", name = mailbox, from = self.name })
end

function _M:pub(topic, text)
  return self:rpc({ op = "pub", name = topic, from = self.name, text = text or "" })
end

function _M:sub(prefix)
  return self:rpc({ op = "sub", name = prefix or "", from = self.name })
end

function _M:ping()
  return self:rpc({ op = "ping" })
end

function _M:ready(service)
  return self:rpc({ op = "ready", name = service, from = self.name })
end

function _M:next_id()
  self.seq = self.seq + 1
  return self.prefix .. "-" .. self.seq
end

-- req requires a correlation id; the hub rejects the frame without one.
function _M:req(service, text, id)
  return self:rpc({ op = "req", name = service, from = self.name, text = text or "", id = id or self:next_id() })
end

function _M:rep(id, text, name)
  return self:rpc({ op = "rep", id = id, name = name or "", from = self.name, text = text or "" })
end

function _M:reserve(mailbox, lease)
  return self:rpc({ op = "reserve", name = mailbox, from = self.name, lease = lease or 60 })
end

function _M:ack(delivery)
  return self:rpc({ op = "ack", delivery = delivery })
end

function _M:nack(delivery)
  return self:rpc({ op = "nack", delivery = delivery })
end

function _M:rpc(obj)
  local ok, err = write_frame(self.sock, obj)
  if not ok then
    return nil, err
  end
  local f, rerr = self:recv()
  if not f then
    return nil, rerr
  end
  if f.op == "err" then
    return nil, f.error or "zmqcat error"
  end
  return f
end

function _M:recv()
  while true do
    local f, err = read_frame(self.sock)
    if not f then
      return nil, err
    end
    if f.op == "ping" then
      write_frame(self.sock, { op = "pong", id = f.id or "" })
    else
      return f
    end
  end
end

function _M:close()
  if self.sock then
    self.sock:close()
  end
end

return _M
